tests/core/data/test_network.py

```diff
@@ -0,0 +1,334 @@
+
+def test_init_raise_if_edges_not_unique():
+    """in_edges and out_edges must be unique."""
+    node = random_node((2, 2))
+    with pytest.raises(ValueError):
+       network = Network([node[0]], [node[0]])
+
+def test_non_dangling_in_or_out_raises():
+    """Input edges for in_edges and out_edges must be dangling."""
+    node = random_node((2,))
+    node2 = random_node((2,))
+    node[0] ^ node2[0]
+
+    with pytest.raises(ValueError):
+       network = Network([node[0]], [])
+
+    with pytest.raises(ValueError):
+       network = Network([], [node[0]])
+
+def test_nodes_dont_include_edges_raises():
+    """Included nodes do not have the passed in_edges or out_edges."""
+    node = random_node((2,2))
+    node2 = random_node((2,2))
+    with pytest.raises(ValueError):
+        network = Network(node[0:1], node[1:], nodes=[node2], copy=False)
+
+def test_non_in_out_edges_dangling_raises():
+    """Included nodes can not have dangling edges that are not in_edges or
+    out_edges
+    """
+    node = random_node((2,2,2))
+    with pytest.raises(ValueError):
+        network = Network([node[0]], [node[1]], copy=False)
+
+def test_scalar_nodes():
+    """It is possible to include scalar nodes as long as they are not the only
+    nodes.
+    """
+    array = np.random.random()
+    node_scalar = tn.Node(array)
+    node = random_node((2,2))
+    network = Network([node[0]], [node[1]], nodes=[node, node_scalar],
+                      copy=False)
+    assert network.nodes <= set([node, node_scalar])
+
+def test_scalar_network():
+    """Test that networks that contains only a scalar node can be created.
+    """
+    node = random_node(())
+    network = Network([], [], nodes=[node])
+    np.testing.assert_allclose(node.tensor, network.nodes.pop().tensor)
+
+def test_scalar_network_contraction():
+    node = random_node(())
+    network = Network([], [], nodes=[node])
+    np.testing.assert_allclose(node.tensor, network.to_array())
+
+def test_empty_rises():
+    with pytest.raises(ValueError):
+       network = Network([], [], nodes=[])
+
+    with pytest.raises(ValueError):
+       network = Network([], [], nodes=None)
+
+def test_copy_in_init():
+    node = random_node((2,2))
+    network = Network(node[0:1], node[1:], copy=False)
+    assert node is network.nodes.pop()
+    network = Network(node[0:1], node[1:], copy=True)
+    network_node = network.nodes.pop()
+    assert network_node is not node
+    assert network_node.tensor is node.tensor
+
+def test_copy():
+    node = random_node((2,2))
+    network = Network(node[0:1], node[1:], copy=False)
+
+    copied_network = network.copy()
+    assert copied_network.in_edges[0] is not network.in_edges[0]
+    assert copied_network.out_edges[0] is not network.out_edges[0]
+
+    assert copied_network.nodes is not network.nodes
+    node = network.nodes.pop()
+    copied_node = copied_network.nodes.pop()
+    assert node is not copied_node
+    assert node.tensor is copied_node.tensor
```

Comment:
We should assert that the copied networks are equal, even if it's just with `network.to_array()` and `copied_network.to_array()` for now.

Reply:
I can definitely include that assertion but I do not think it is necessary here. Most of the networks in this test file have a single node and that is to facilitate these kind of comparisons. For instance, here, by doing `assert node.tensor is copied_node.tensor` we already check that both the copied and original nodes have the exact same information. In fact, they contain the _same_ tensor which means they save in memory when copying the nodes.